src/algebra/group/to_additive.lean

```diff
@@ -104,6 +140,23 @@ meta def reorder_attr : user_attribute (name_map $ list ℕ) (list ℕ) :=
 
 end extra_attributes
 
+/--
+Find the first argument of `nm` that has a multiplicative type-class on it.
+Returns 1 if there are no types with a multiplicative class as arguments.
+E.g. `prod.group` returns 1, and `pi.has_one` returns 2.
+-/
+meta def first_multiplicative_arg (nm : name) : tactic ℕ := do
+  d ← get_decl nm,
+  let (es, _) := d.type.pi_binders,
+  l ← es.mmap_with_index $ λ n bi, do {
+    let tgt := bi.type.pi_codomain,
+    let n_bi := bi.type.pi_binders.fst.length,
+    tt ← has_attribute' `to_additive tgt.get_app_fn.const_name | return none,
+    let n2 := tgt.get_app_args.head.get_app_fn.match_var.map $ λ m, n + n_bi - m,
+    return $ n2 },
+  let l := l.reduce_option,
```

Comment:
Is there any chance that `l.reduce_option` will return an empty list? e.g. `l` is empty or full of `none`? Please add an assertion here (and right after `let (es, _) := d.type.pi_binders`) if not or an error message if so.

Reply:
Yes, if we have a declaration that has no multiplicative type-class arguments (this happens in most classes, like `has_mul` and `group`).